cmd/shipperctl/cmd/backup/restore.go

```diff
@@ -89,47 +88,40 @@ func runRestoreCommand(cmd *cobra.Command, args []string) error {
 	return nil
 }
 
-func restore(releasesPerApplications []shipperBackupObject, kubeClient kubernetes.Interface, shipperClient versioned.Interface) error {
-	var errList []string
+func restore(releasesPerApplications []shipperBackupObject, kubeClient kubernetes.Interface, shipperClient shipperclientset.Interface) error {
 	for _, obj := range releasesPerApplications {
 		// apply application
 		if err := applyApplication(kubeClient, shipperClient, obj.Application); err != nil {
-			errList = append(errList, err.Error())
-			continue
+			return err
```

Comment:
Wouldn't this stop the `restore` on every error? (as oppose to stop only on the first error)?

Reply:
If `applyApplication` fails, `restore` will throw the same error, which will be the returned error by the command.
The first error encountered by the restoring process will stop the command entirely.
I think this is better since, if the restore only partially worked, I would delete all restored objects and try again

Reply:
Yes, totally understand that.

But I thought that we were going for a different approach:
- Either fail if the incumbent release fails, or
- Keep going and collect of the errors if the "history" release fails

I think in this case we could be in a situation where an old release comes back to life 🧟

Reply:
I'm a bit confused
> I think in this case we could be in a situation where an old release comes back to life 🧟

this case == panic out on every error?

> But I thought that we were going for a different approach:

I would assume that in case I am trying to restore a backup and I have failed to restore some of the history releases (the only error I have seen so far is "already exists", so I don't know which errors can occur), I wouldn't be happy with this situation and I would delete everything and restart the restore from scratch (knowing that shipper is done so I'm not loosing any new data).

I guess I should clarify on the usage of `restore` that shipper must be down during restore...

Reply:
Just to wrap this up for future reference, we agree that any time there is an error we bail because we want to restore _all_ the data, history is as important as the incumbent release. If there is an error the operator should clean up the partially restored objects and retry.

Reply:
The `restore` command needs two pre-steps:
- Shipper must be down
- Shipper objects must be deleted from the cluster

Crashing with an "Application already exists" message to the operator restoring the backup could be confusing on what would be the next thing the operator must do.
Adding to that the stress level of an operator restoring a Shipper backup after a procedural error was made, and this could be slowing down a lot the process of recovery.

What I suggest is to actually implement verifiers of those pre-steps at the beginning of the `restore` command.
- "check if Shipper is up, crash if not with a very informative message on what action should be done by the operator"
- "check if there is some Shipper object in the cluster, crash if not with a very informative message on what action should be done by the operator"

WDYGT?

Reply:
@ArchangelX360  @nodo 
For the second point: if some shipper objects were found in the cluster, should the command offer to delete them? Or only bail out with "please clean" message?

Reply:
Bailing out I think, keep the command dumb is usually a good thing in under stress operation workflow.

Ideally, the operator should follow a list of instructions that mentions:
1. take down Shipper by running x, y, z
2. delete Shipper objects by running x, y, z
3. restore backup by running `shipperctl backup restore`
These checks I am suggesting here is to prevent a "I skipped an instruction" situation.

In the future could decide to make step one and 2 part of `shipperctl backup restore` in the future to have a fully automated procedure with manual confirmation prompt. This does not mean that the command will be smart, it won't recover by itself, but taking the correct step in the same order, which is less confusing.

I think for the sake of this first MR, we can stay simple and not implement step 1 and 2, just checking that they are done by the operator.

Reply:
I would rather not delete the object during restore, I think it might be a very unexpected behaviour.

Reply:
Even with a prompt to ask about it before?
And what about scaling down shipper? Should the command (ask to do it and then) do it? Or only verify?

Reply:
Yeah, just verify and just bail out with an error in both cases. 

The rationale behind is
- deleting objects is easy to do with `kubectl` anyway
- adds complexity to these commands
- During a restore if you see a unwanted object or a problem you might want to take a step back and check more manual checks

Reply:
expected output for the 2 checks:

```
## If shipper deoployment is not scaled down:
$ shipperctl backup restore
Making sure shipper is down... Error! shipper deployment has 1 replicas. Scale it down first with `kubectl -n shipper-system patch deploy shipper --type=merge -p '{"spec":{"replicas":1}}'`

## If there are shipper objects in the cluster:
$ shipperctl backup restore
Making sure shipper is down... done
Making sure there are no Shipper objects lying around... Error! found Shipper objects:
 - Applications: 3
 - Releases: 3
 - InstallationTargets: 3
 - TrafficTargets: 3
 - CapacityTargets: 3
delete them first with `kubectl delete app,rel,it,tt,ct --all-namespaces --all`
```